File: src/main/java/io/scriptor/loader/ClassTree.java
package io.scriptor.loader;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Collections;
import java.util.Iterator;
import java.util.NoSuchElementException;

public record ClassTree(@NotNull ResourceClassLoader classLoader, @NotNull String packageName)
        implements ITree<Class<?>> {

    private record ClassEntry(@NotNull Class<?> value) implements ITree<Class<?>> {

        @Override
        public Iterator<ITree<Class<?>>> children() {
            return Collections.emptyIterator();
        }

        @Override
        public boolean hasValue() {
            return true;
        }
    }

    @Override
    public Iterator<ITree<Class<?>>> children() {
        return new Iterator<>() {

            private final Iterator<IPackageEntry> iterator;

            {
                try {
                    iterator = classLoader.getClassPackage(packageName).iterator();
                } catch (final IOException | ClassNotFoundException e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public boolean hasNext() {
                return iterator.hasNext();
            }

            @Override
            public ITree<Class<?>> next() {
                final var entry = iterator.next();
                if (entry instanceof IPackageEntry.ClassEntry(Class<?> type)) {
                    return new ClassEntry(type);
                }
                if (entry instanceof IPackageEntry.FutureEntry(ResourceClassLoader loader, String name)) {
                    return new ClassTree(loader, name);
                }
                throw new NoSuchElementException();
            }
        };
    }

    @Override
    public Class<?> value() {
        return null;
    }

    @Override
    public boolean hasValue() {
        return false;
    }
}

File: src/main/java/io/scriptor/loader/ClassTree.kt
package io.scriptor.loader

import java.io.IOException
import java.util.*

data class ClassTree(val classLoader: ResourceClassLoader, val packageName: String) : ITree<Class<*>> {

    private data class ClassEntry(override val value: Class<*>) : ITree<Class<*>> {

        override val children: MutableIterator<ITree<Class<*>>>
            get() = Collections.emptyIterator()

        override fun hasValue(): Boolean {
            return true
        }
    }

    override val children
        get(): MutableIterator<ITree<Class<*>>> {
            return object : MutableIterator<ITree<Class<*>>> {
                private val iterator: MutableIterator<IPackageEntry>

                init {
                    try {
                        iterator = classLoader.getClassPackage(packageName).iterator()
                    } catch (e: IOException) {
                        throw RuntimeException(e)
                    } catch (e: ClassNotFoundException) {
                        throw RuntimeException(e)
                    }
                }

                override fun hasNext(): Boolean {
                    return iterator.hasNext()
                }

                override fun next(): ITree<Class<*>> {
                    val entry = iterator.next()
                    if (entry is IPackageEntry.ClassEntry) {
                        return ClassEntry(entry.type)
                    }
                    if (entry is IPackageEntry.FutureEntry) {
                        return ClassTree(entry.classLoader, entry.packageName)
                    }
                    throw NoSuchElementException()
                }

                override fun remove() {
                    throw UnsupportedOperationException()
                }
            }
        }

    override val value = null

    override fun hasValue(): Boolean {
        return false
    }
}
